contrib/aws/internal/sns/sns.go

```diff
@@ -0,0 +1,101 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+package sns
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/sns"
+	"github.com/aws/aws-sdk-go-v2/service/sns/types"
+	"github.com/aws/smithy-go/middleware"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+const (
+	datadogKey           = "_datadog"
+	maxMessageAttributes = 10
+)
+
+type messageCarrier map[string]string
+
+func (carrier messageCarrier) Set(key, val string) {
+	carrier[key] = val
+}
+
+func EnrichOperation(ctx context.Context, in middleware.InitializeInput, operation string) {
+	switch operation {
+	case "Publish":
+		handlePublish(ctx, in)
+	case "PublishBatch":
+		handlePublishBatch(ctx, in)
+	}
+}
+
+func handlePublish(ctx context.Context, in middleware.InitializeInput) {
+	params, ok := in.Parameters.(*sns.PublishInput)
+	if !ok {
+		log.Debug("Unable to read PublishInput params")
+		return
+	}
+
+	if params.MessageAttributes == nil {
+		params.MessageAttributes = make(map[string]types.MessageAttributeValue)
+	}
+
+	injectTraceContext(ctx, params.MessageAttributes)
+}
+
+func handlePublishBatch(ctx context.Context, in middleware.InitializeInput) {
+	params, ok := in.Parameters.(*sns.PublishBatchInput)
+	if !ok {
+		log.Debug("Unable to read PublishBatch params")
+		return
+	}
+
+	for i := range params.PublishBatchRequestEntries {
+		entryPtr := &params.PublishBatchRequestEntries[i]
+		if entryPtr.MessageAttributes == nil {
+			entryPtr.MessageAttributes = make(map[string]types.MessageAttributeValue)
+		}
+		injectTraceContext(ctx, entryPtr.MessageAttributes)
+	}
+}
+
+func injectTraceContext(ctx context.Context, messageAttributes map[string]types.MessageAttributeValue) {
```

Comment:
If this is exactly the same as the sqs inject code, we could combine it into a single location.

Reply:
Hmm, the problem is they use different `messageAttributes` from the AWS SDK. It can be done, but it would require a lot of extra boiler plate like switch statements. Something like this, what do you think?
```
switch ma := messageAttributes.(type) {
	case map[string]sqsTypes.MessageAttributeValue:
		if len(ma) >= maxMessageAttributes {
			log.Info("Cannot inject trace context: message already has maximum allowed attributes")
			return
		}

		ma[datadogKey] = sqsTypes.MessageAttributeValue{
			DataType:    aws.String("String"),
			StringValue: aws.String(string(jsonBytes)),
		}
	case map[string]snsTypes.MessageAttributeValue:
		if len(ma) >= maxMessageAttributes {
			log.Info("Cannot inject trace context: message already has maximum allowed attributes")
			return
		}

		ma[datadogKey] = snsTypes.MessageAttributeValue{
			DataType:    aws.String("String"),
			StringValue: aws.String(string(jsonBytes)),
		}
	}
```

Reply:
Hmm, could be best to just keep them separate then. I'll let you make the call.